### PR TITLE
MTDSA-17184 - create checkpoint endpoint

### PR DESCRIPTION
**Pull request by jeremystone**
- add HATEOAS support
- add HATEOAS to endpoint
- Add (temporary?) validation to detect missing nino

### CODE REVIEW

File: app/api/connectors/BaseDownstreamConnector.scala [hunk 1]
```diff
@@ -69,6 +79,16 @@ trait BaseDownstreamConnector extends Logging {
     http.put(url(downstreamUri)).withBody(Json.toJson(body)).execute
   }
 
+  def putEmpty[Resp](downstreamUri: DownstreamUri[Resp])(implicit
```
Comment (harry-ritchie) [inline app/api/connectors/BaseDownstreamConnector.scala:82]: For which endpoint do we need a PUT with no request body?

File: app/uk/gov/hmrc/mtdsatestsupportapi/models/hateoas/CheckpointHateoasLinks.scala [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+  val SELF = "self"
+
+  private def baseUrl(appConfig: AppConfig) =
+    s"/${appConfig.apiGatewayContext}/vendor-state/checkpoints"
+
+  def createCheckpoint(appConfig: AppConfig, nino: Nino): Link =
+    Link(href = s"${baseUrl(appConfig)}/?nino=${nino.value}", method = POST, rel = "create-checkpoint")
+
+  def restoreCheckpoint(appConfig: AppConfig, checkpointId: CheckpointId): Link =
```
Comment (harry-ritchie) [inline app/uk/gov/hmrc/mtdsatestsupportapi/models/hateoas/CheckpointHateoasLinks.scala:35]: I think the restore url should be `href = s"${baseUrl(appConfig)}/restore/${checkpointId.value}"`, or does the order not matter here?

File: app/api/connectors/BaseDownstreamConnector.scala (hunk 1 above)
Comment (jeremystone) [reply app/api/connectors/BaseDownstreamConnector.scala:82]: There was talk of making the restore a PUT. But this is for completeness.

File: app/uk/gov/hmrc/mtdsatestsupportapi/models/hateoas/CheckpointHateoasLinks.scala (hunk 2 above)
Comment (jeremystone) [reply app/uk/gov/hmrc/mtdsatestsupportapi/models/hateoas/CheckpointHateoasLinks.scala:35]: I think the spec is wrong -- from a REST viewpoint the restore is below the checkpoint's own resource. (I will fix spec). The code agrees with the HATEOAS spec.

**Pull request merged**